Add MAX31865 sensor support, fix MAX31855 sensor

DAVe3283 opened the pull request:
## Description:

### MAX31865

Added support for the MAX31865 RTD-to-Digital Converter to measure PT100 and similar RTDs. Verified with an Adafruit unit (product ID: 3328) and a PT100 probe.

### MAX31855

This was setup for incorrect SPI clock polarity and phase, and would return bad data due to a race condition measuring on the wrong edge (verified with Saleae Logic scope). Selecting the correct configuration fixes that problem.

Re-wrote the decode off the datasheet to handle error states better (sends NaN as an update on failure to read temperature, which shows the value as Unknown in Home Assistant).

Added the *optional* ability to monitor the internal high-precision temperature sensor, which can be nice in some applications.

**Related issue (if applicable):** N/A

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#399

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

DAVe3283 committed:
Add MAX31865 sensor support, fix MAX31855 sensor.

# MAX31865

Added support for the MAX31865 RTD-to-Digital Converter to measure PT100 and
similar RTDs. Verified with an Adafruit unit (product ID: 3328) and a PT100
probe.

# MAX31855

This was setup for incorrect SPI clock polarity and phase, and would return bad
data due to a race condition measuring on the wrong edge (verified with Saleae
Logic scope). Selecting the correct configuration fixes that problem.

Re-wrote the decode off the datasheet to handle error states better (sends NaN
as an update on failure to read temperature, which shows the value as Unknown
in Home Assistant).

Added the *optional* ability to monitor the internal high-precision temperature
sensor, which can be nice in some applications.

DAVe3283 committed:
Merge remote-tracking branch 'upstream/dev' into dev

DAVe3283 committed:
Tests for MAX31855/MAX38165.

DAVe3283 committed:
Update style to match project rules.

Also fix CONF_REFERENCE_RESISTANCE and CONF_REFERENCE_TEMPERATURE being defined
multiple places. Missed this when I added them to const.py.

DAVe3283 committed:
Update style to match project rules.

Pylint line limit 101/100 ("missed it by that much").
Also apparently I can't read and patched the wrong line in max31855.cpp.

sylviamic reviewed (approved):
This looks nice, and appears to work with my own MAX31865 sensor. Great work! Looking forward to seeing this merged.

DAVe3283 committed:
Minor string/style cleanup.

There was a copy-paste leftover in max31855.cpp and max31865/sensor.py had
unnecessary whitespace.

DAVe3283 committed:
Improve MAX31865 fault detection and logging.

Log levels are more in-line with the documented descriptions.

Fault detection code is improved. A transient fault between reads is still
reported, but now only faults *during* a read cause the sensor to fail and
return NAN ("unknown" in Home Assistant).

DAVe3283 committed:
Update style to match project rules.

I just now realized the .clang-format and pylintrc files are included. D'oh!

DAVe3283 commented:
I am using this code successfully to control a chilled water bath. The ESP32 and MAX31865 (and MAX31855 before it) are located in the same housing as the chiller compressor, which is a very electrically noisy environment. These latest changes to fault detection have the sensor operating very smoothly, handling the occasional overvoltage fault when the compressor kicks on.

I am very confident in this code for both the MAX31855 and MAX31865, as I have now run both for several days with no software issues. I can report back again in another few days if desired, but I feel this code is ready to merge.

OttoWinter commented on esphome/components/max31855/max31855.cpp:
```diff
@@ -22,63 +23,84 @@ void MAX31855Sensor::setup() {
   this->read_array(data, 4);
+  this->disable();
+
+  const uint32_t mem(data[0] << 24 | data[1] << 16 | data[2] << 8 | data[3] << 0);
```
```suggestion
  const uint32_t mem = data[0] << 24 | data[1] << 16 | data[2] << 8 | data[3] << 0;
```

I'm not sure why you used the `()` constructor syntax for declaring variables - but IMO it's much harder to read. I strongly recommend using the equals syntax.

(Applies in other places too).

DAVe3283 replied:
Just habit from other projects. I will happily update to match the project style later tonight.

OttoWinter commented on esphome/components/max31855/max31855.cpp:
```diff
@@ -22,63 +23,84 @@ void MAX31855Sensor::setup() {
-  // val is 14 bits of signed temperature data followed by 2 bits of status flags
-  int16_t val = data[1] | data[0] << 8;
+  // Verify we got data
+  if (mem && mem != 0xFFFFFFFF) {
```
```suggestion
  if (mem != 0 && mem != 0xFFFFFFFF) {
```

Style suggestion.

OttoWinter commented on esphome/components/max31855/sensor.py:
```diff
@@ -1,14 +1,16 @@
 
 CONFIG_SCHEMA = sensor.sensor_schema(UNIT_CELSIUS, ICON_THERMOMETER, 1).extend({
     cv.GenerateID(): cv.declare_id(MAX31855Sensor),
+    cv.Optional(CONF_REFERENCE_TEMPERATURE):
```
Usually, I would not allow this syntax (sensor schema within sensor schema) - should be base schema with two sensor schemas in it (see also dht). But it's ok here for avoiding a breaking change 👍

DAVe3283 committed:
MAX31855 & MAX31865 code style alignment.

@OttoWinter caught some style mismatches, updated to match project better.

DAVe3283 committed:
Fix a lost '\' in max31865/sensor.py.

DAVe3283 commented:
The MAX31865 sensor has been working great all week, enabling a climate component to run the chiller.
![Soda Fountain](https://user-images.githubusercontent.com/6894502/68270890-251d8f80-001c-11ea-8ee5-a4bbe41f4050.png)
The output of this sensor is so good I've got the water bath very close to freezing with no issues, even in the noisy electrical environment it has to deal with.
![Water Temperature](https://user-images.githubusercontent.com/6894502/68270718-7b3e0300-001b-11ea-9a36-3381b8d29c16.png)
